### PR TITLE
Traces fetching: new "split and join" strategy

### PR DESCRIPTION
**Pull request by jotak**
Fixes https://github.com/kiali/kiali/issues/3311

Capture d’écran de 2020-12-09 09-07-33

### CODE REVIEW

**Comment by jotak** [comment]
Some remarks:

- This strategy works both for traces & spans fetching (so, tracing page and metrics page w/ correlation), and for all target kinds (apps/workloads/services)

- The "split-join" strategy might still not be sufficient to see traces spread all over the time, when a lot of traces are generated. At the moment I have an arbitrary division of 10 intervals slices on which traces are going to be fetched. These divisions might still be too coarse-grained in high volume to really see a smooth distribution of traces. We could increase the number of intervals to fetch, but it increases the number of queries performed against Jaeger, which could cause performance penalty I fear. Perhaps we could do that if we adopt the jaeger protobuf client? (which we should probably do at some point anyway, I guess...)

- I removed the previous "find relevant spans" algorithm that we had (it was used only when fetching spans for correlation with metrics, not for the tracing page). These two algorithms are somehow competing in providing more spans when the query limit was reached. It can be debated, perhaps we can also imagine a mix of the two algorithms...

**Comment by jotak** [comment]
By the way, one convenient side effect of this change is that the "adjust time" checkbox is now pretty useless since data will often be spread all over the timeline. One display setting that we could remove.

**Comment by lucasponce** [comment]
Thanks @jotak, I'm catching up with some pending reviews but this one is interesting, thanks for implementing.
I'll review it locally.

**Review by hhovsepy — APPROVED** [review]
Regression tested for Traces and Metrics.

File: business/jaeger.go [hunk 1]
```diff
@@ -99,7 +102,23 @@ func (in *JaegerService) GetAppTraces(ns, app string, query models.TracingQuery)
 	if err != nil {
 		return nil, err
 	}
-	return client.GetAppTraces(ns, app, query)
+	r, err := client.GetAppTraces(ns, app, query)
+	if err != nil {
+		return nil, err
+	}
+	if len(r.Data) == query.Limit {
```
Comment (lucasponce) [inline business/jaeger.go:109]: Ok, so, if I understand the implementation, first it fetches until reach the limit (i.e. 20 traces).
If limit is reached, then it performs the split and try to query 10 times over the given interval, dividing the limit (20 traces) with the slices (10).

Ok, so, in that case, when I have a query with 20 limit, it'll return 40 results.

Isn't it ?

Or perhaps we can just query the very same limit on the slices, and allow the user to personalize the slices as a parameter.

In any case, it sounds ok as a first iteration, I know it's not an easy topic.

Comment (jotak) [reply business/jaeger.go:109]: > Ok, so, if I understand the implementation, first it fetches until reach the limit (i.e. 20 traces).
> If limit is reached, then it performs the split and try to query 10 times over the given interval, dividing the limit (20 traces) with the slices (10).
> 
> Ok, so, in that case, when I have a query with 20 limit, it'll return 40 results.
> 
> Isn't it ?

Exactly

> 
> Or perhaps we can just query the very same limit on the slices, and allow the user to personalize the slices as a parameter.

Yes, perhaps. That should go with changes in the UI too, then. Perhaps we can think about that as part of our discussion about the toolbar redesign? ( here https://github.com/kiali/kiali/issues/2776 )

> 
> In any case, it sounds ok as a first iteration, I know it's not an easy topic.

Comment (lucasponce) [reply business/jaeger.go:109]: > Yes, perhaps. That should go with changes in the UI too, then. Perhaps we can think about that as part of our discussion about the toolbar redesign? ( here #2776 )

Totally, it makes sense.

**Review by lucasponce — APPROVED** [review]

**Pull request merged**